redash/serializers.py

```diff
@@ -45,9 +46,17 @@ def public_dashboard(dashboard):
     widget_list = (models.Widget.query
                    .filter(models.Widget.dashboard_id == dashboard.id)
                    .outerjoin(models.Visualization)
-                   .outerjoin(models.Query))
+                   .outerjoin(models.Query)
+                   #.with_entities(models.Widget, models.Query.latest_query_data_id)
+                   )
+
     widgets = {w.id: public_widget(w) for w in widget_list}
 
+    # make sure all widgets' query's have a last_query_data_id that is not null so public dashboards work
+    for w in widgets:
+        if not hasattr(w, 'latest_query_data'):
+            models.Query.query.filter(models.Query.id == widgets[w]['visualization']['query']['id'])
```

Comment:
I mean that `filter` returns a new query object, but you need to do something with it, like `.all()` or `.first()`, to execute it.